src/ApplicationInsights.ServiceFabric.Native/Net45/MethodNameProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿namespace Microsoft.ApplicationInsights.ServiceFabric.Remoting.Activities
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+
+    /// <summary>
+    /// Provider for looking up method names by an interface id and a method id. It has methods allowing proxies
+    /// or service objects to be examined, and its interfaces and methods will be added to the map. The ids are
+    /// calculated using the same logic as service fabric runtime, so the method ids used in service remoting can
+    /// be mapped back to method names. 
+    /// </summary>
+    internal class MethodNameProvider : IMethodNameProvider
+    {
+        private Dictionary<int, Dictionary<int, string>> idToMethodNameMap;
+        private HashSet<Type> processedTypes;
+        private object thisLock;
+
+        /// <summary>
+        /// Instantiates the <see cref="MethodNameProvider"/> with the specified remoting factory and retrysettings.
+        /// </summary>
+        public MethodNameProvider()
+        {
+            this.idToMethodNameMap = new Dictionary<int, Dictionary<int, string>>();
+            processedTypes = new HashSet<Type>();
+            thisLock = new object();
+        }
+
+        /// <summary>
+        /// Look up the given interface id and method id and returns the method name.
+        /// This only works for interface types for which it had created proxies.
+        /// </summary>
+        /// <param name="interfaceId"></param>
+        /// <param name="methodId"></param>
+        /// <returns></returns>
+        public string GetMethodName(int interfaceId, int methodId)
+        {
+            lock (thisLock)
```

Comment:
I assume if idToMethodNameMap type is changed to ConcurrentDictionary, lock is not longer needed.

Reply:
Removed the lock by switching to ConcurrentDictionary.